File: algorithms/0406.queue_reconstruction_by_height/queue_reconstruction_by_height.rs
impl Solution {
    pub fn reconstruct_queue(people: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        if people.is_empty() {
            return people;
        }
        let mut people = people;
        people.sort_by(|a, b| {
            if a[1].cmp(&b[1]) != std::cmp::Ordering::Equal {
                return a[1].cmp(&b[1]);
            }
            a[0].cmp(&b[0])
        });
        // println!("people = {:?}", people);
        for index in 1..people.len() {
            let cur_person = &people[index];
            let mut count = 0;
            for (i, person) in people.iter().enumerate() {
                if person[0] >= cur_person[0] {
                    count += 1;
                }
                if count > cur_person[1] && i < index {
                    let e = people.remove(index);
                    people.insert(i, e);
                    break;
                }
            }
        }
        people
    }
}

File: algorithms/0406.queue_reconstruction_by_height/queue_reconstruction_by_height.cpp
class Solution {
public:
    vector<vector<int>> reconstructQueue(vector<vector<int>>& people) {
        auto cmp = [](const auto &a, const auto &b) {
            if (a[0] == b[0]) { return a[1] < b[1]; }
            return a[0] > b[0];
        };
        sort(people.begin(), people.end(), cmp);
        list<vector<int>> queue;
        for (const auto &p: people) {
            int position = p[1];
            auto it = queue.begin();
            while (position--) {
                ++it;
            }
            queue.insert(it, p);
        }
        return vector<vector<int>>(queue.begin(), queue.end());
    }
};